### DISCUSSION TITLE
How can I query a relation within Matrix via GraphQL

### DISCUSSION THREAD

**Discussion by tremby** [craftcms/cms#14546] [Q&A] [answered]
I have products, and I have recipes.

When I view a product, I want to show a recipe which mentions that product as one of its ingredients.

The ingredients are currently modelled as a Matrix field `recipeIngredients`, with one block type, `ingredient`. That block type has a few fields, one of which is a regular Craft entry relation field, `linkedProduct`.

How can I construct my GQL query to fetch recipe entries which refer to a given product in (at least) one of those matrix blocks? I have the relevant product's site ID, section, ID, and URI.

To complicate it, the recipe type has another relation field which points to products, `relatedProducts`, and I do *not* want to take any relations here into account (they are not necessarily ingredients). To be clear, if the product is related both here *and* in an ingredient, that's fine, it should be included. But not if it's only mentioned in `relatedProducts`, and not in `recipeIngredients`.

Do I want `relatedTo`? `relatedToEntries`? The documentation on both is confusing. The in-GQL documentation is unhelpful since it says one takes `QueryArgument` which just says it's a string or integer or boolean with no clue on what specifically they do (and searching for QueryArgument in the Craft docs does not turn up anything useful), and the other takes `EntryCriteriaInput`, which has a ton of fields, none of which is `linkedProduct` or `recipeIngredients`.

**Comment by brandonkelly** [comment, marked as answer]
There’s not a great way to do this right now.

You could preload all the recipe + ingredient relations up front as part of your build pipeline, and reference that locally, if that’s an option.

Otherwise, if you really need to find related recipes on the fly, you could add a “Combined Products” relation field to the Recipes section, and set its value dynamically when recipes are saved.

```php
use craft\base\Event;
use craft\elements\Entry;
use craft\events\ModelEvent;
use craft\fields\BaseRelationField;

Event::on(Entry::class, Entry::EVENT_AFTER_PROPAGATE, function(ModelEvent $event) {
    /** @var Entry $entry */
    $entry = $event->sender;

    // is this entry in the Recipes section?
    if ($entry->section?->handle === 'recipes') {
        // get all the products related to its ingredients
        $productIds = [];
        foreach ($entry->ingredients->with('ingredient:product') as $ingredient) {
            foreach ($ingredient->product as $product) {
                $productIds[] = $product->id;
            }
        }

        /** @var BaseRelationField $field */
        $field = Craft::$app->fields->getFieldByHandle('combinedProducts', 'global');
        Craft::$app->relations->saveRelations($field, $entry, $productIds);
    }
});
```

In that code I’m assuming:

- the section handle is `recipes`
- the Matrix field handle is `ingredient`
- the nested product relation field handle is `product`
- the newly-created “Combined Products” relation field handle is `combinedProducts`

(Adjust accordingly.)

You can even hide that “Combined Products” field from view for authors, by setting an admin-only condition on the field within the Recipe entry type’s field layout.

With that in place, you could grab the related recipes using the `relatedTo` argument, or `relatedToEntries` (or whatever the appropriate variant is).

**Reply by tremby** [reply]
That's helpful; thanks.

> You could preload all the recipe + ingredient relations up front as part of your build pipeline

There's something like 12,000 products so I don't think we can do this.

The other option looks reasonable; thank you for the example code.

> With that in place, you could grab the related recipes using the `relatedTo` argument, or `relatedToEntries` (or whatever the appropriate variant is).

Can you help with this part, please? As far as I can tell from the docs, `relatedTo` lets me query for those matching one or more IDs, but doesn't allow me to tell it a specific relation field to look at. As I mentioned before, there's another field which also references products, which mustn't be allowed to poison the results.

Meanwhile, `relatedToEntries` looks to have a more sophisticated input form but I see no "field" or "via" or "through" or "relation" field here either.

In case it's not clear, I need to do something akin to

```graphql
query {
  recipeEntries(
    relatedToEntries: {
      id: $myId
      viaField: "referencedIngredients"
    }
  ) {
    title
  }
}
```

But I see nothing in the GQL API to fill the role of `viaField` above.

**Reply by brandonkelly** [reply]
Sorry, I should have mentioned, you can target that specific field using the field’s own argument:

```graphql
query {
  recipeEntries(
    referencedIngredients: $myId
  ) {
    title
  }
}
```

**Reply by tremby** [reply]
Oh wow, that's simpler than I expected.

I have it all working. Thanks so much for your help.

**Reply by brandonkelly** [reply]
Glad to hear!
